src/lib/swapper/swappers/CowSwapper/cowExecuteTrade/cowExecuteTrade.ts

```diff
@@ -22,57 +20,37 @@ import { cowService } from 'lib/swapper/swappers/CowSwapper/utils/cowService'
 import type { CowSwapOrder } from 'lib/swapper/swappers/CowSwapper/utils/helpers/helpers'
 import {
   domain,
+  getCowswapNetwork,
   getNowPlusThirtyMinutesTimestamp,
   hashOrder,
 } from 'lib/swapper/swappers/CowSwapper/utils/helpers/helpers'
 
-export async function cowExecuteTrade(
-  { apiUrl, adapter }: CowSwapperDeps,
-  { trade, wallet }: ExecuteTradeInput<KnownChainIds.EthereumMainnet>,
+import { validateExecuteTrade } from '../utils/validator'
+
+export async function cowExecuteTrade<T extends CowChainId>(
+  { baseUrl: apiUrl, adapter }: CowSwapperDeps,
+  { trade, wallet }: ExecuteTradeInput<T>,
 ): Promise<Result<TradeResult, SwapErrorRight>> {
-  const cowTrade = trade as CowTrade<KnownChainIds.EthereumMainnet>
+  const cowTrade = trade as CowTrade<T>
+  const network = getCowswapNetwork(adapter)
   const {
-    sellAsset,
-    buyAsset,
     feeAmountInSellTokenCryptoBaseUnit: feeAmountInSellToken,
     sellAmountDeductFeeCryptoBaseUnit: sellAmountWithoutFee,
+    buyAsset,
     accountNumber,
     id,
     minimumBuyAmountAfterFeesCryptoBaseUnit,
   } = cowTrade
 
-  const { assetReference: sellAssetErc20Address, assetNamespace: sellAssetNamespace } = fromAssetId(
-    sellAsset.assetId,
-  )
-  const { assetReference: buyAssetErc20Address, chainId: buyAssetChainId } = fromAssetId(
-    buyAsset.assetId,
-  )
+  const maybeValidTradePair = validateExecuteTrade(cowTrade)
+  maybeValidTradePair.isErr() && Err(maybeValidTradePair.unwrapErr())
 
-  if (sellAssetNamespace !== 'erc20') {
-    return Err(
-      makeSwapErrorRight({
-        message: '[cowExecuteTrade] - Sell asset needs to be ERC-20 to use CowSwap',
-        code: SwapErrorType.UNSUPPORTED_PAIR,
-        details: { sellAssetNamespace },
-      }),
-    )
-  }
-
-  if (buyAssetChainId !== KnownChainIds.EthereumMainnet) {
-    return Err(
-      makeSwapErrorRight({
-        message: '[cowExecuteTrade] - Buy asset needs to be on ETH mainnet to use CowSwap',
-        code: SwapErrorType.UNSUPPORTED_PAIR,
-        details: { buyAssetChainId },
-      }),
-    )
-  }
+  const { sellAssetAddress, buyAssetAddress } = maybeValidTradePair.unwrap()
 
-  const buyToken =
-    buyAsset.assetId !== ethAssetId ? buyAssetErc20Address : COW_SWAP_ETH_MARKER_ADDRESS
+  const buyToken = buyAsset.assetId !== ethAssetId ? buyAssetAddress : COW_SWAP_ETH_MARKER_ADDRESS
```

Comment:
That doesn't generalize to multiple chains (will only work for Ethereum currently), check against the `feeAssetId` of the current chain using `adapter.getFeeAssetId()`

Reply:
yeah I'm aware, just wasn't sure on how to address this